Add support for importing generic cloud images

cevich opened the pull request:
***DEPENDS ON:***
https://github.com/containers/automation_images/pull/178

Downstream CI needs dictate early testing on new Fedora releases at the
Beta stage or earlier.  Unfortunately at the time of this commit, the
Fedora cloud-sig does not provide ready-made beta images in AWS EC2.

Add a new image-build stage called `import_images`, that takes care of
bringing the generic images into AWS.  Update the `base_images` stage
such that it will locate and use the imported images.  Unify all the
build-stage wrapper scripts into a single/common `ci/make.sh` wrapper.

Finally, make a few semi-related comment/readability improvements in
various places.

---


Workaround packer amazon_import post-processor bug


Ref: https://github.com/hashicorp/packer-plugin-amazon/issues
     number 264

Implement a rough and ready set of make targets as a stand-in for
the broken packer amazon_import post-processor.  This works, but
is fairly fragile:  There are several failure modes which cannot
easily be checked automatically.  For example, the s3 upload can
silently fail and/or store corrupted data.

When the packer bug is resolved, this commit can most probably be
safely  reverted.  This should make the process entirely automated
(via the `automation_images` PR workflow), possibly with only
minimal updates.

Add a `import_images/README.md` reference for the semi-manual
process.

cevich commented:
[Yay progress](https://cirrus-ci.com/task/5205715336822784), I decoded that message and it's actually surprisingly helpful.  I think I can fix the "permission denied" with a simple IAM policy update.

cevich commented:
`status code: 400, request id: a4757af1-9640-4e69-989e-4b9722860e3d`

Blah! That's not helpful at all Amazon :astonished:

cevich commented:
Found the error in AWS CloudTrail:

```JSON
    "errorCode": "Client.InvalidParameter",
    "errorMessage": "The service role vmimport provided does not exist or does not have sufficient permissions",
```

lsm5 commented on import_images/handle_image.sh:
```diff
@@ -0,0 +1,43 @@
+#
+# The first argument is the file path and name for the downloaded image,
+# the second argument is the image download URL (ending in a filename).
+# The third argument is a the download URL for a checksum file containing
```
Nit: pick one: `a the`.

cevich replied:
Whut? Is that TOO specific or something? :grinning: 

Thanks, I'll fix it.

lsm5 commented on Makefile:
```diff
@@ -13,6 +13,23 @@ export_full = $(eval export $(if $(call err_if_empty,$(1)),$(1)))
+export FEDORA_IMAGE_URL = https://dl.fedoraproject.org/pub/fedora/linux/development/37/Cloud/x86_64/images/Fedora-Cloud-Base-37-20220830.n.0.x86_64.qcow2
+export FEDORA_CSUM_URL = https://dl.fedoraproject.org/pub/fedora/linux/development/37/Cloud/x86_64/images/Fedora-Cloud-37-x86_64-20220830.n.0-CHECKSUM
+export FEDORA_ARM64_IMAGE_URL = https://dl.fedoraproject.org/pub/fedora/linux/development/37/Cloud/aarch64/images/Fedora-Cloud-Base-37-20220830.n.0.aarch64.qcow2
+export FEDORA_ARM64_CSUM_URL = https://dl.fedoraproject.org/pub/fedora/linux/development/37/Cloud/aarch64/images/Fedora-Cloud-37-aarch64-20220830.n.0-CHECKSUM
```
no biggie, but may be convenient to have these URLs crafted using something like: $(BASE_URL)/$(FEDORA_RELEASE)/$(FOO)/$(BAR)

cevich replied:
It's a good suggestion. But would make copy-pasting the link harder.    Actually I was thinking about if/how they could be looked up based on the release number.  Especially the development case, the filenames change daily, sometimes multiple times per day.  The release filenames less so, but I have seen them change too.

cevich replied:
https://github.com/containers/automation_images/issues/190

lsm5 reviewed (approved):
LGTM (to the best of my experience :D). Will it be possible to see the results of image import in github comments prior to merge, or would that be only a post-merge thing?

cevich commented:
> LGTM (to the best of my experience :D). Will it be possible to see the results of image import in github comments prior to merge, or would that be only a post-merge thing?

This is mostly all brand-new to me, so out experience level is pretty close with this.  Also why it's taking me so long to get it working :disappointed: 

Assuming I can get it working, this will allow us to support rawhide and development images in AWS and/or GCP.  It removes all dependence on the fedora cloud-sig publishing images into AWS for us to consume.  It modifies the build workflow to go like:

import_images -> base_images -> cache_images

So yes, everything will show up in the handy table :grin:

cevich commented:
Some slight progress, now the error (from AWS) says `failed with status message: ClientError: Unknown OS / Missing OS files., error: ResourceNotReady: failed waiting for successful resource state`

cevich commented:
and `retry count exhausted. Last err: InvalidParameter: Parameter architecture = arm64 has an invalid format.`

cevich commented:
Ref: https://www.packer.io/plugins/post-processors/amazon

cevich commented:
Force-push: Trying a different disk-format (VHDX) instead of raw.

cevich commented:
(In packer's defense, a LONG-standing AWS documentation bug is likely to blame)

cevich commented:
Ref: https://github.com/hashicorp/packer/issues/ number 11970

cevich commented:
This has a good chance of breaking / being unreliable.  In manual testing, it was about 50% working.  I believe the issue is, uploading to s3 seems completely unverified.  The smallest (otherwise) undetectable network "blip" seems to corrupt the image and cause the import-service to barf.  I don't see a way around this other than running the import manually :cry:

cevich commented:
For frickin' frog snacks it's finally working.  Sheesh.  I thought AWS was a PITA ***before*** this PR.  It's reached an entirely new level now :sob:

cevich commented:
@lsm5 this is a two-week old festering feline hairball of a PR.  PTAL.  Especially the new `README.md`, I'd feel more comfortable knowing my bad grammar makes sense to someone else.  Better yet, give the process a try and confirm it "*works*" for > "just me".  Where "*works*" is used quite loosely, since it's actually pretty unreliable until you run it about 13 times :grin:

cevich commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  Retry `make import images`.
+* If import was successful but AMI tagging failed, manually add
+  the required tags to AMI: `automation=false` and `Name=<name>-i${IMG_SFX}`.
+  Where `<name> is `fedora-aws` or `fedora-aws-arm64`.
```
```suggestion
  Where `<name>` is `fedora-aws` or `fedora-aws-arm64`.
```

cevich commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  changes were needed, run `make clean`.
+  Retry `make import images`.
+* If snapshot import fails with a `Disk validation failed` error,
+  remove the uploaded object from S3.
```
```suggestion
```

lsm5 commented:
@cevich I'll try out the changes in README.md.

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+   $ make image_builder_debug \
+         IMG_SFX=$(date +%s) \
+         GAC_FILEPATH=/dev/null \
+         AWS_SHARED_CREDENTIALS_FILE=/path/to/.aws/credentials \
```
```suggestion
         AWS_SHARED_CREDENTIALS_FILE=/path/to/.aws/credentials
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+* If checksum validation fails,
+  double-check the URL values.  If
+  changes made run `make clean`.
+  Retry `make import images`.
```
```suggestion
  Retry `make import_images`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+* If s3 upload fails
+  double-check the URL values.  If
+  changes were needed, run `make clean`.
+  Retry `make import images`.
```
```suggestion
  Retry `make import_images`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  Retry `make import images`.
+* If snapshot import fails with a `Disk validation failed` error,
+  remove the uploaded object from S3.
+  Retry `make import images`.
```
```suggestion
  Retry `make import_images`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  remove the uploaded object from S3.
+  Retry `make import images`.
+* If snapshot import fails with an error, find them in EC2 and delete them.
+  Retry `make import images`.
```
```suggestion
  Retry `make import_images`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+* If snapshot import fails with an error, find them in EC2 and delete them.
+  Retry `make import images`.
+* If AMI registration fails, remove any conflicting AMIs and snapshots.
+  Retry `make import images`.
```
```suggestion
  Retry `make import_images`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+This list is not exhaustive, and only represents common/likely failures.
+Normally there is no need to exit the build container.
+
+* If image download fails, double-check the URL values, rune `make clean`
```
```suggestion
* If image download fails, double-check the URL values, run `make clean`
```

cevich replied:
But it's all magic, so actually a rune is required here :rofl:

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  and retry.
+* If checksum validation fails,
+  double-check the URL values.  If
+  changes made run `make clean`.
```
```suggestion
  changes made, run `make clean`.
```

lsm5 commented on import_images/README.md:
```diff
@@ -0,0 +1,94 @@
+  double-check the URL values.  If
+  changes made run `make clean`.
+  Retry `make import images`.
+* If s3 upload fails
```
```suggestion
* If s3 upload fails,
```

lsm5 commented on import_images/handle_image.sh:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+# This script is intended to be run by packer, usage under any other
+# environment may behave badly.  It's purpose is to download a VM
```
```suggestion
# environment may behave badly. It's purpose is to download a VM
```

lsm5 commented on import_images/handle_image.sh:
```diff
@@ -0,0 +1,45 @@
+
+# This script is intended to be run by packer, usage under any other
+# environment may behave badly.  It's purpose is to download a VM
+# image and a checksum file.  Verify the image's checksum matches.
```
```suggestion
# image and a checksum file. Verify the image's checksum matches.
```

lsm5 commented on import_images/wait_import_task.sh:
```diff
@@ -0,0 +1,73 @@
+
+# This script is intended to be called by the main Makefile
+# to wait for and confirm successful import and conversion
+# of an uploaded image object from S3 into EC2.  It expects
```
```suggestion
# of an uploaded image object from S3 into EC2. It expects
```

lsm5 commented on import_images/wait_import_task.sh:
```diff
@@ -0,0 +1,73 @@
+# first argument.
+#
+# If the import is successful, the snapshot ID is written
+# to stdout.  Otherwise, all output goes to stderr, and
```
```suggestion
# to stdout. Otherwise, all output goes to stderr, and
```

lsm5 commented on import_images/wait_import_task.sh:
```diff
@@ -0,0 +1,73 @@
+#
+# If the import is successful, the snapshot ID is written
+# to stdout.  Otherwise, all output goes to stderr, and
+# the script exits non-zero on failure or timeout.  On
```
```suggestion
# the script exits non-zero on failure or timeout. On
```

lsm5 commented:
import_images failed with:

```
You must specify a region. You can also configure your region by running "aws configure".
make: *** [Makefile:266: /tmp/automation_images_tmp/fedora-aws-1662750353.import_task_id] Error 255
```

Is this expected to be grabbed from the user's aws config? Or is it preset in automation_images?

cevich commented:
> Is this expected to be grabbed from the user's aws config? Or is it preset in automation_images?

Ahh right, okay.  It's expected to be in the config file.  However we only support one region: `us-east-1`.  So I should probably just force that on all the `aws` command-lines.

cevich commented:
BTW: Thanks @lsm5 for such a through look-though, I really appreciate the oversight.

lsm5 commented:
> BTW: Thanks @lsm5 for such a through look-though, I really appreciate the oversight.

not really that thorough TBH :D , but happy to be useful. I will give this another run once you have the `us-east-1` change pushed.

cevich commented:
>  I will give this another run once you have the us-east-1 change pushed.

Thanks.  Just testing them out now, along with a change to the way we upload the image to S3.  I found a way to get S3 to verify a checksum on upload and retry on mismatch.  That will make the upload more reliable.

cevich commented:
@lsm5 okay, I addressed all your concerns, fixed the region problem, and added support for verifying the MD5 of uploaded images before conversion.  Testing still shows the upload is somewhat unreliable and a few times the AWS vmimport service just timed out or puked on otherwise valid content.  It's almost as if they didn't want anybody to automate this :cry: 

Note: I rebased this onto #192 so it can all merge at once (assuming it's ready).

lsm5 commented:
ack, retrying...

lsm5 reviewed (approved):
```
Image import(s) successful.
############################################################
Please update Makefile value:

    FEDORA_IMPORT_IMG_SFX = 1663007754
############################################################
rm /tmp/automation_images_tmp/fedora-aws-1663007754.uploaded /tmp/automation_images_tmp/fedora-aws-arm64-1663007754.import_task_id /tmp/automation_images_tmp/fedora-aws-1663007754.ami.name /tmp/automation_images_tmp/fedora-aws-arm64-1663007754.ami.name /tmp/automation_images_tmp/fedora-aws-arm64-1663007754.uploaded /tmp/automation_images_tmp/fedora-aws-1663007754.md5 /tmp/automation_images_tmp/fedora-aws-arm64-1663007754.md5 /tmp/automation_images_tmp/fedora-aws-1663007754.import_task_id
```

/lgtm

lsm5 commented:
/hold

cevich commented:
There we go, okay perfect.  Thanks again @lsm5
